File: src/main/java/catchla/yep/util/OkMediaDownloader.java
package catchla.yep.util;

import android.support.annotation.NonNull;

import org.mariotaku.mediaviewer.library.CacheDownloadLoader;
import org.mariotaku.mediaviewer.library.MediaDownloader;

import java.io.IOException;
import java.io.InputStream;

import catchla.yep.model.CacheMetadata;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import okhttp3.ResponseBody;

/**
 * Created by mariotaku on 16/1/20.
 */
public class OkMediaDownloader implements MediaDownloader {
    private final OkHttpClient client;

    public OkMediaDownloader(final OkHttpClient client) {
        this.client = client;
    }

    @NonNull
    @Override
    public CacheDownloadLoader.DownloadResult get(@NonNull final String url, final Object extra) throws IOException {
        final Request.Builder builder = new Request.Builder();
        builder.url(url);
        Response response = client.newCall(builder.build()).execute();
        if (!response.isSuccessful()) {
            response.body().close();
            throw new IOException("Unable to get " + url + ": " + response.code());
        }
        final ResponseBody body = response.body();
        CacheMetadata metadata = new CacheMetadata();
        metadata.setContentType(body.contentType().toString());
        return new OkDownloadResult(body, metadata);
    }


    private static class OkDownloadResult implements CacheDownloadLoader.DownloadResult {
        private final ResponseBody mBody;
        private final CacheMetadata mMetadata;

        public OkDownloadResult(ResponseBody body, CacheMetadata metadata) {
            mBody = body;
            mMetadata = metadata;
        }

        @Override
        public void close() throws IOException {
            mBody.close();
        }

        @Override
        public long getLength() throws IOException {
            return mBody.contentLength();
        }

        @Override
        public InputStream getStream() throws IOException {
            return mBody.byteStream();
        }

        @Override
        public byte[] getExtra() {
            if (mMetadata == null) return null;
            final String serialize = JsonSerializer.serialize(mMetadata, CacheMetadata.class);
            if (serialize == null) return null;
            return serialize.getBytes();
        }

        @Override
        protected void finalize() throws Throwable {
            close();
            super.finalize();
        }
    }
}

File: src/main/java/catchla/yep/util/OkMediaDownloader.kt
package catchla.yep.util

import catchla.yep.model.CacheMetadata
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.ResponseBody
import org.mariotaku.mediaviewer.library.CacheDownloadLoader
import org.mariotaku.mediaviewer.library.MediaDownloader
import java.io.IOException
import java.io.InputStream

/**
 * Created by mariotaku on 16/1/20.
 */
class OkMediaDownloader(private val client: OkHttpClient) : MediaDownloader {

    @Throws(IOException::class)
    override fun get(url: String, extra: Any?): CacheDownloadLoader.DownloadResult {
        val builder = Request.Builder()
        builder.url(url)
        val response = client.newCall(builder.build()).execute()
        if (!response.isSuccessful) {
            response.body().close()
            throw IOException("Unable to get " + url + ": " + response.code())
        }
        val body = response.body()
        val metadata = CacheMetadata()
        metadata.contentType = body.contentType().toString()
        return OkDownloadResult(body, metadata)
    }


    data class OkDownloadResult(
            private val body: ResponseBody,
            private val metadata: CacheMetadata?
    ) : CacheDownloadLoader.DownloadResult {

        @Throws(IOException::class)
        override fun close() {
            body.close()
        }

        @Throws(IOException::class)
        override fun getLength(): Long {
            return body.contentLength()
        }

        @Throws(IOException::class)
        override fun getStream(): InputStream {
            return body.byteStream()
        }

        override fun getExtra(): ByteArray? {
            if (metadata == null) return null
            val serialize = JsonSerializer.serialize(metadata, CacheMetadata::class.java) ?: return null
            return serialize.toByteArray()
        }

        protected fun finalize() {
            close()
        }
    }
}
